compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeParamAnalyzer.java

```diff
@@ -768,39 +839,68 @@ private BRecordType getMatchingRecordBoundType(BRecordType expType, SymbolEnv en
 
         LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
         for (BField expField : expType.fields.values()) {
+            BType type = expField.type;
+            BType matchingBoundType = getMatchingBoundType(type, env, resolvedTypes);
+            if (!hasDifferentType && type != matchingBoundType) {
+                hasDifferentType = true;
+            }
             BField field = new BField(expField.name, expField.pos,
                                       new BVarSymbol(0, expField.name, env.enclPkg.packageID,
-                                                     getMatchingBoundType(expField.type, env, resolvedTypes),
-                                                     env.scope.owner, expField.pos, VIRTUAL));
+                                              matchingBoundType, env.scope.owner, expField.pos, VIRTUAL));
             fields.put(field.name.value, field);
             recordSymbol.scope.define(expField.name, field.symbol);
         }
 
         BRecordType bRecordType = new BRecordType(recordSymbol);
         bRecordType.fields = fields;
         recordSymbol.type = bRecordType;
+        bRecordType.flags = expType.flags;
 
         if (expType.sealed) {
             bRecordType.sealed = true;
         }
-        bRecordType.restFieldType = getMatchingBoundType(expType.restFieldType, env, resolvedTypes);
+        BType restFieldType = expType.restFieldType;
+        bRecordType.restFieldType = getMatchingBoundType(restFieldType, env, resolvedTypes);
+        if (!hasDifferentType && restFieldType != bRecordType.restFieldType) {
```

Comment:
We can get the matching bound type for the `restFieldType`  before creating the `BRecordType` in  L854, right? If so we don't have to create the `bRecordType` if it's the same type yeah?